pkg/controller/clusteroperation/clusteroperation.go

```diff
@@ -0,0 +1,158 @@
+	if err != nil {
+		return err
+	}
+
+	// Create a source for watching ClusterOperation events.
+	src := &source.Kind{Type: &corev1alpha1.ClusterOperation{}}
+	// Create a handler for handling events from ClusterOperation resource.
+	h := &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &corev1alpha1.ClusterOperation{},
+	}
+	pred := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			logrus.Info("CREATE")
+			err := c.updateClusterOperationStatus(e.Object)
+			if err != nil {
+				//TODO: handle
+			}
+			//TODO: sdk call
+			return true
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			logrus.Info("UPDATE")
+			//TODO: sdk call
+			return e.MetaOld.GetGeneration() != e.MetaNew.GetGeneration()
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			logrus.Info("DELETE")
+			//TODO: sdk call
+			// Evaluates to false if the object has been confirmed deleted.
+			return !e.DeleteStateUnknown
+		},
+	}
+	// Watch for ClusterOperation events.
+	return ctrl.Watch(src, h, pred)
+}
+
+func (c *Controller) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) updateClusterOperationStatus(object runtime.Object) error {
+	instance := &corev1alpha1.ClusterOperation{}
+	//TODO: get object, update status
+	return c.client.Status().Update(context.TODO(), instance)
+}
+
+// createCRD creates the CRD for ClusterOperation object
+func (c *Controller) createCRD() error {
+
+	crdName := fmt.Sprintf("%s.%s", corev1alpha1.ClusterOperationResourcePlural, corev1alpha1.SchemeGroupVersion.Group)
+
+	crd := &apiextensionsv1beta1.CustomResourceDefinition{
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name: crdName,
+		},
+		Spec: apiextensionsv1beta1.CustomResourceDefinitionSpec{
+			Group:   corev1alpha1.SchemeGroupVersion.Group,
+			Version: corev1alpha1.SchemeGroupVersion.Version,
+			Scope:   apiextensionsv1beta1.NamespaceScoped,
+			Names: apiextensionsv1beta1.CustomResourceDefinitionNames{
+				Singular:   corev1alpha1.ClusterOperationResourceName,
+				Plural:     corev1alpha1.ClusterOperationResourcePlural,
+				Kind:       reflect.TypeOf(corev1alpha1.ClusterOperation{}).Name(),
+				ShortNames: []string{corev1alpha1.ClusterOperationShortName},
+			},
+			Subresources: &apiextensionsv1beta1.CustomResourceSubresources{
+				Status: &apiextensionsv1beta1.CustomResourceSubresourceStatus{},
+			},
+		},
+	}
+	err := k8s.Instance().RegisterCRD(crd)
+	if err != nil && !errors.IsAlreadyExists(err) {
+		return err
+	}
+
+	err = k8s.Instance().ValidateCRD(k8s.CustomResource{
+		Name:       crd.ObjectMeta.Name,
+		Plural:     crd.Spec.Names.Plural,
+		Group:      crd.Spec.Group,
+		Version:    crd.Spec.Version,
+		Scope:      crd.Spec.Scope,
+		Kind:       crd.Spec.Names.Kind,
+		ShortNames: crd.Spec.Names.ShortNames,
+	}, validateCRDTimeout, validateCRDInterval)
+	return err
+}
+
+func WatchForCompletion() {
```

Comment:
remove until there is an actual use case. The portworx sdk stuff won't happen here but rather inside the volume driver interface. The controller loop needs to be very generic.